### DISCUSSION TITLE
[Feature Proposal]: Support for multiple color schemes

### DISCUSSION THREAD

**Discussion by afuno** [tailwindlabs/tailwindcss#1505] [Ideas]
Hello.

I wanted to offer you my idea for implementing support for several color schemes. In my opinion, support for at least light and dark designs is necessary for many applications.

Now in the configuration file there is a `theme` object, which describes absolutely everything. Including colors:

```js
theme: {
  colors: {
    body: '#ffffff',
    primary: '#003366',
    secondary: '#cccccc',
  }
},
```

I suggest by default to do this:

```js
theme: {
  colors: {
    default: {
      body: '#ffffff',
      primary: '#003366',
      secondary: '#cccccc',
    }
  }
},
```

This implies that these colors will be the default. If it is necessary to add, for example, a dark version, then we do it like this:

```js
theme: {
  colors: {
    default: {
      body: '#ffffff',
      primary: '#003366',
      secondary: '#cccccc',
    },

    dark: {
      body: '#000000',
      primary: '#003366',
      secondary: '#333333',
    }
  }
},
```

Now for the initialization in the template. Initialize the color scheme, which by default, makes no sense. But for everyone else, I suggest specifying the name of the schema, for example, in the tag `body`:

```html
<body class="dark">
  <!-- ... -->
</body>
```

Or:

```html
<body class="xs:dark">
  <!-- ... -->
</body>
```

What do you think about this?

---

_P.S. I know about https://github.com/hawezo/tailwindcss-theming._

**Comment by afuno** [comment]
In version 3 a similar implementation method appeared: https://github.com/innocenzi/tailwindcss-theming/blob/master/docs/quick-start.md#usage

**Comment by estevanmaito** [comment]
I follow with my opinion. Please, correct me if I'm wrong. I would also like to hear your thoughts on it.

I think what makes current _theming_ with Tailwind great is the low barrier and naturality that someone can start, whatever their background. It says: _that's how we do colors here_

In this proposed approach, in my opinion, it would create a barrier: _forget the colors you know, now you need to give them names_

Take this image from [Material Design guidelines on dark themes](https://material.io/design/color/dark-theme.html):

![](https://lh3.googleusercontent.com/pgTa3nA1suThryEsoRszChgrhhEFisnZ8HWA8RYC0LQUzCPXKHq-lO00N0vUNnSxklor7fNBOzAuNqPnN4jRAy4QPcTDI2gjBAC96w=w800-v0)

In the light theme, we can see that the chart and navigation are _primary_, while in the dark theme, the navigation turns black and the chart adopts what looks like the _primary_ for dark.

Another point, in the light theme, the background and cards are white, we could call them _background_. But in the dark variant, there's actually a distinction between them. It would not be possible to theme it without creating a dark utility that applies only on dark.

Giving names to colors would mean that at some point we would be naming `primary`. But the navigation should receive another color on dark, so let's create the `primary-navigation` or, create an utility like `dark:bg-gray-800`, just to apply to elements that would need this kind of adaptation.

## The solution I see

Maintain the current approach, but offer theming options. Add a property to the config (could be inside `theme`, in the root...), let's call it `themeVariants` (horrible name), where you pass an array containing the names of the themes you pretend to use, in my case would be:

```js
themeVariants: ['dark', 'neon']
```
It would create variants for every utility like `dark:bg-green-100` etc., and would expect a parent element containing some sort of identifier class generated by `themeVariants`, let's say `.dark-theme`.

Positive points:

- If you know how to use Tailwind now, you will know how to create themes;
- It allows nested themes: you could use a dark section inside a light theme and vice-versa;
- There's no friction. No need to touch configurations or giving things names;
- If you need a custom color scheme, you only need to define it once (and not for every theme variant).

Negative points:

- As there's no way of knowing what kind of theme variant is being created, JavaScript would be needed to get user's theme preference (`prefers-color-scheme`). In your approach you could simply duplicate the `dark` scheme inside `@media (prefers-color-scheme: dark)` and it would be solved (talking about auto applying themes, as both would need JS to deliberately change theme).

An excerpt from [Building dark mode on Stack Overflow](https://stackoverflow.blog/2020/03/31/building-dark-mode-on-stack-overflow/) that supports _the Tailwind approach_:

> To offer complete flexibility, Stacks provides atomic color classes that are only applied when dark mode is enabled. You can read about Stacks CSS design choices at length at my personal portfolio. By adding `.d:bg-green-100` to an element, our engineers and designers can say “In dark mode, apply a background of green 100.” Additional conditional classes allow us to drop borders, swap backgrounds, or change text colors in dark mode. Steve Schoger’s got a [really great tweet](https://twitter.com/steveschoger/status/1151160261170126850?s=20) demonstrating the customizations that are sometimes required for dark modes. I’ve taken lots of inspiration from **Tailwind**.

This plugins does this (edit: but only for dark mode): https://github.com/ChanceArthur/tailwindcss-dark-mode

**Comment by estevanmaito** [comment]
I created a plugin based on this idea: https://github.com/estevanmaito/tailwindcss-multi-theme#tailwind-css-multi-theme

You can see it live here: https://tailwindcss-multi-theme.now.sh/

**Comment by paulpopus** [comment]
I want to support for this as well. I've been doing some research on several approaches involving tailwind but none of them seem ideal, I'll write below my findings.

## Using the @ screens method

In your config you'll configure something like this
```js
screens: {
     ...
     light: { raw: "(prefers-color-scheme: light)" },
     dark: { raw: "(prefers-color-scheme: dark)" },
},
```
The result is that you can then do something like `light:text-black dark:text-white` and you can even combine them with variants `light:hover:text-blue` which is incredibly useful.

The downside of this method is that you rely on browser support for `(prefers-color-scheme: dark)` and that media query is not something we can change in the front end based on user preference per website, so it ignores any manual overrides you may provide and won't work with adding classes to the body or html elements.

## Using a custom variant

There's a few modules that handle this, including [tailwind-css-multi-theme](https://github.com/estevanmaito/tailwindcss-multi-theme#tailwind-css-multi-theme) mentioned above and [tailwindcss-dark-mode](https://github.com/ChanceArthur/tailwindcss-dark-mode) but writing your own isn't terribly difficult. These modules just generate variants for each theme variant for each possible CSS variant like hover or focus.

Here is an example code

```js
plugins: [
   function({ addVariant, e, theme }) {
      const themeVariants = {
        'dark': 'dark-mode',
        'light': 'light-mode',
      };
      for (const [mode, target] of Object.entries(themeVariants)) {
        addVariant(mode, ({modifySelectors, separator}) => {
          modifySelectors(({className}) => {
            return `.${target} .${e(`${mode}${separator}${className}`)}`;
          });
        });
        let hoverMode = mode + '-hover';
        addVariant(hoverMode, ({modifySelectors, separator}) => {
          modifySelectors(({className}) => {
            return `.${target} .${e(`${hoverMode}${separator}${className}`)}:hover`;
          });
        });
      }
    },
]
```
then in your variants for your components you need to add the respective variants, eg `textColor: ['dark', 'dark-hover']`

The variant method works pretty well to get what you want and you can easily extend it to use the theme configuration. However the amount of variants you suddenly have to account for can grow quickly, assuming we have a dark and a light mode at the minimum you can generate up to 8 variants just to cover basic pseudo selectors.

## My proposed solution

This is something I've tried to do but I didn't manage to write a plugin that maintains any existing variants, but the idea is to  create a component that wraps the rules and the selectors with a different target.

So from this configuration

```js
themeVariants: {
  'dark': 'dark-mode',
}
```

You can write `class="text-black dark:text-white dark:hover:text-red"` where the expected output CSS looks something like

```css
.text-black {
  color: black;
}

.dark-mode {
  .dark\:text-white {
    color: text-white;
  }

  .dark\:hover\:text-white:hover {
    color: text-red;
  }
}
```

I think it's also very important to note that color schemes for a website can go beyond just light and dark, there's high contrast modes, specific colorblind modes, prefers reduced motion and many others.

Now, all that said. This issue would also solve this problem https://github.com/tailwindlabs/tailwindcss/issues/1373 which might be a better approach than trying to shoehorn another solution into tailwind.

**Comment by estevanmaito** [comment]
@paulpopus Did I miss how your solution is any different than what I proposed? Your solution is also exactly what `tailwindcss-multi-theme` does:

1. You define a variant name `themeVariants: ['dark']` (or more for multi themes)
2. It encapsulates all variants under `.theme-<your variant name>`
3. You just use, eg. `dark:<whatever>`

**Comment by paulpopus** [comment]
@estevanmaito no, I was just re-iterating the available options and saying that both of those tailwind plugins do the same thing as the code I posted. I was saying that I think it's a bad solution to require so many variants and ideally we will be able to stack variants in the future, as per https://github.com/tailwindlabs/tailwindcss/issues/1373

**Comment by danmatthews** [comment]
Hi All, stumbled into this thread as i came here to raise a proposition for adding this to core as a way to be able to add a `high-contrast` theme that could be triggered by adding a class to the body using Javascript.

The reason being, i'm currently working on a client site that requires a high-contrast version, so i'm currently using the dark mode system as a way to enable that.

Now that we have the JIT, is there potentially a use-case for generating any unrecognised variants as a parent class in the same way the dark mode stuff can work with body classes? Though this might lead to pollution of selectors.

Otherwise, the `themeVariants` stuff looks like the right route to go, although possibly only for custom themes OTHER than dark themes? so as not to change the config syntax again for 2.0 projects.

**Comment by christianwiedemann** [comment]
Hi all, with JIT compile the samples above did not work for me. After some resarch I found a way which work with JIT.

```
const { updateLastClasses, transformAllSelectors } = require('tailwindcss/lib/util/pluginUtils');
const prefixSelector = require('tailwindcss/lib/util/prefixSelector').default;
module.exports = {
  mode: 'jit',
.....
  plugins: [
    forms,
    typography,
    plugin(function({ addVariant }) {
      const variants = ['red', 'light'];
      variants.forEach((variant) => {
        addVariant(
          variant,
          transformAllSelectors((selector) => {
            const variantSelector = updateLastClasses(selector, (className) => {
              return `${variant}:${className}`;
            });
            if (variantSelector === selector) {
              return null;
            }
            const darkSelector = prefixSelector('', `.${variant}`);
            return `${darkSelector} ${variantSelector}`;
          })
        );
      });
    }),
  ],
}

```

**Comment by justingolden21** [comment]
Is there any way to add multiple themes (with or without third party libraries) where I can simply use all of Tailwind's colors (50,100,200,300...) as I already use them with `gray` in my project but with the other grays? I wouldn't want to write something like `dark:` for each of these in every location across thousands of lines, when all I need is the ability to swap `gray-400` with `slate-400` etc. everywhere.

**Reply by 7iomka** [reply]
did you find a solution?
